libraries/AP_Torqeedo/AP_Torqeedo.cpp

```diff
@@ -0,0 +1,310 @@
+#define TORQEEDO_PACKET_FOOTER      0xAD    // communication packer footer
+
+extern const AP_HAL::HAL& hal;
+
+// parameters
+const AP_Param::GroupInfo AP_Torqeedo::var_info[] = {
+
+    // @Param: ENABLE
+    // @DisplayName: Torqeedo enable
+    // @Description: Torqeedo enable
+    // @Values: 0:Disabled, 1:Enabled
+    // @User: Standard
+    // @RebootRequired: True
+    AP_GROUPINFO_FLAGS("ENABLE", 1, AP_Torqeedo, _enable, 0, AP_PARAM_FLAG_ENABLE),
+
+    // @Param: ONOFF_PIN
+    // @DisplayName: Torqeedo ON/Off pin
+    // @Description: Pin number connected to Torqeedo's on/off pin. -1 to disable turning motor on/off from autopilot
+    // @Values: -1:Disabled,50:AUX1,51:AUX2,52:AUX3,53:AUX4,54:AUX5,55:AUX6
+    // @User: Standard
+    // @RebootRequired: True
+    AP_GROUPINFO("ONOFF_PIN", 2, AP_Torqeedo, _pin_onoff, -1),
+
+    // @Param: DE_PIN
+    // @DisplayName: Torqeedo DE pin
+    // @Description: Pin number connected to RS485 to Serial converter's DE pin. -1 to disable sending commands to motor
+    // @Values: -1:Disabled,50:AUX1,51:AUX2,52:AUX3,53:AUX4,54:AUX5,55:AUX6
+    // @User: Standard
+    // @RebootRequired: True
+    AP_GROUPINFO("DE_PIN", 3, AP_Torqeedo, _pin_de, -1),
+
+    AP_GROUPEND
+};
+
+AP_Torqeedo::AP_Torqeedo()
+{
+    _singleton = this;
+    AP_Param::setup_object_defaults(this, var_info);
+}
+
+// initialise driver
+void AP_Torqeedo::init()
+{
+    // only init once
+    if (_initialised) {
+        return;
+    }
+
+    // create background thread to process serial input and output
+    if (!hal.scheduler->thread_create(FUNCTOR_BIND_MEMBER(&AP_Torqeedo::loop, void), "torqeedo", 4096, AP_HAL::Scheduler::PRIORITY_RCOUT, 1)) {
+        return;
+    }
+}
+
+// initialise serial port and gpio pins (run from background thread)
+void AP_Torqeedo::init_internals()
+{
+    // find serial driver and initialise
+    const AP_SerialManager &serial_manager = AP::serialmanager();
+    _uart = serial_manager.find_serial(AP_SerialManager::SerialProtocol_Torqeedo, 0);
+    if (_uart == nullptr) {
+        return;
+    }
+    _uart->begin(TORQEEDO_SERIAL_BAUD);
+    _uart->set_flow_control(AP_HAL::UARTDriver::FLOW_CONTROL_DISABLE);
+    _uart->set_unbuffered_writes(true);
+
+    // initialise onoff pin and set for 0.5 sec to turn on motor
+    if (_pin_onoff > -1) {
+        hal.gpio->pinMode(_pin_onoff, HAL_GPIO_OUTPUT);
+        hal.gpio->write(_pin_onoff, 1);
+
+        // delay 0.5 sec and then unset pin
+        hal.scheduler->delay(500);
+        hal.gpio->write(_pin_onoff, 0);
+    }
+
+    // initialise RS485 DE pin (when high, allows send to motor)
+    if (_pin_de > -1) {
+        hal.gpio->pinMode(_pin_de, HAL_GPIO_OUTPUT);
+        hal.gpio->write(_pin_de, 0);
+    }
+
+    _initialised = true;
+}
+
+// consume incoming messages from motor, reply with latest motor speed
+// runs in background thread
+void AP_Torqeedo::loop()
+{
+    // initialisation
+    init_internals();
+    if (!_initialised) {
```

Comment:
This doesn't look right; `init_internals` is definitely not idempotent.
```suggestion
    if (!_initialised) {
        init_internals();
```

Reply:
ok txs.  I've changed init_internals to return a boolean and renamed "loop" to "thread_main".